### DISCUSSION TITLE
Usage of wave chatbot_card with pandas-ai

### DISCUSSION THREAD

**Discussion by KarstenSchnappauf** [h2oai/wave#2049] [Q&A] [answered]
Hi,

I want to use to the chatbot_card from wave in combination with pandas-ai for interactions with a Dataframe. 

At first, my question is, whether the syntax of the following code snippet ist right? because I don't see the output of the LLM in the UI, just in terminal. 

The second question is, whether the chat_card is able to visualize charts/ Images? Because pandas-ai is able to create from a prompt charts from the dataframe.

```py
query = "SELECT * FROM mytable"
df = pd.read_sql_query(query, conn)


@app('/demo')
async def serve(q: Q):
    if not q.client.initialized:
        q.page['chat'] = ui.chatbot_card(box='1 1 5 5', data=data(fields='content from_user', t='list'), name='chatbot')
        q.client.initialized = True

    if q.args.chatbot:
        user_message = q.args.chatbot
        q.page['chat'].data += [[user_message, True]]
        try:
            result = pandas_ai(df, prompt=user_message)
            q.page['chat'].data += [[result, False]]
        except Exception as e:
            q.page['chat'].data += [[str(e), False]]

    await q.page.save()
```

Thank you very much for your answer in advance.

BG,
Karsten

**Comment by mturoci** [comment]
Hi @KarstenSchnappauf.

> whether the syntax of the following code snippet ist right

It's not. Your code has extra brackets. Instead of `q.page['chat'].data += [[user_message, True]]`, `q.page['chat'].data += [user_message, True]` should be used

>  whether the chat_card is able to visualize charts/ Images?

Yes. The chatbot is able to interpret markdown, which supports both images and embedded HTML.

**Reply by KarstenSchnappauf** [reply]
Thank you very much @mturoci,

wie I use your syntax, the whole chatbot_card is disappearing after I entered the prompt. 


`@app('/demo')
async def serve(q: Q):
    if not q.client.initialized:
        q.page['chat'] = ui.chatbot_card(box='1 1 5 5', data=data(fields='content from_user', t='list'), name='chatbot')
        q.client.initialized = True

    if q.args.chatbot:
        user_message = q.args.chatbot
        q.page['chat'].data += [user_message, True]
        try:
            result = pandas_ai(df, prompt=user_message)
            q.page['chat'].data += [result, False]
        except Exception as e:
            q.page['chat'].data += [str(e), False]

    await q.page.save()`

Terminal output: 
INFO:     127.0.0.1:51210 - "POST / HTTP/1.1" 200 OK
2023/06/29 10:58:48 * /62e261b8-88e8-4686-bdd5-83649e686cb4 {"d":[{"k":"chat data __append__","v":["wieviel spalten hat die Tabelle mytable?",true]},{"k":"chat data __append__","v":[14,false]}]}

**Reply by mturoci** [reply, marked as answer]
The `result` in your case is int, rather than string. `q.page['chat'].data += [str(result), False]` should solve it. Can you give it a try and let me know?

**Reply by KarstenSchnappauf** [reply]
works fine now, except that the charts are displayed in an extra window by python and not by the chatbot_card

Thank you very much @mturoci

**Reply by mturoci** [reply]
> charts are displayed in an extra window by python

Hm.. they would need to be part of the `result` to be rendered within the chatbot.

Thanks for the feedback @KarstenSchnappauf! Will add the string conversion directly into Wave to make it less confusing :)

**Reply by KarstenSchnappauf** [reply]
when you have time, you could try it by your own. Currently I don't have a solution for visualizing the charts in the chatbot_card
